[mono][wasm] Allow AOTing methods with catch/filter clauses.

vargaz opened the pull request:
This works as follows:
* During EH, when a catch clause is found in an AOTed method, the
  EH state is saved into TLS and a c++ exception is thrown.
* The C++ exception is caught by the AOTed method, and the landing
  pad code calls mono_llvm_resume_exception_il_state ().
* That call will run the catch clause and the rest of the method
  code using the interpreter, storing the possible return value
  back into the AOTed method's stack frame.
* After the call, the method skips the rest of its code, and
  returns immediately to its caller.

vargaz committed:
[mono][wasm] Allow AOTing methods with catch clauses.

This works as follows:
* During EH, when a catch clause is found in an AOTed method, the
  EH state is saved into TLS and a c++ exception is thrown.
* The C++ exception is caught by the AOTed method, and the landing
  pad code calls mono_llvm_resume_exception_il_state ().
* That call will run the catch clause and the rest of the method
  code using the interpreter, storing the possible return value
  back into the AOTed method's stack frame.
* After the call, the method skips the rest of its code, and
  returns immediately to its caller.

vargaz commented:
/azp run runtime-extra-platforms

vargaz commented:
Running the console benchmark:

BEFORE:
```
|        Exceptions, NoExceptionHandling |     0.4793us |
|                   Exceptions, TryCatch |     0.4888us |
|              Exceptions, TryCatchThrow |     0.0146ms |
|             Exceptions, TryCatchFilter |     0.4176us |
|       Exceptions, TryCatchFilterInline |     0.2308us |
|        Exceptions, TryCatchFilterThrow |     0.0169ms |
| Exceptions, TryCatchFilterThrowApplies |     0.0170ms |
|         Json, non-ASCII text serialize |     0.8488ms |
|       Json, non-ASCII text deserialize |     1.8116ms |
|                  Json, small serialize |     0.0936ms |
|                Json, small deserialize |     0.1772ms |
|                  Json, large serialize |    16.0582ms |
|                Json, large deserialize |    50.0947ms |
```
AFTER:
```
|        Exceptions, NoExceptionHandling |     0.2430us |
|                   Exceptions, TryCatch |     0.3731us |
|              Exceptions, TryCatchThrow |     0.0468ms |
|             Exceptions, TryCatchFilter |     0.5380us |
|       Exceptions, TryCatchFilterInline |     0.3169us |
|        Exceptions, TryCatchFilterThrow |     0.0185ms |
| Exceptions, TryCatchFilterThrowApplies |     0.0177ms |
|         Json, non-ASCII text serialize |     0.9475ms |
|       Json, non-ASCII text deserialize |     1.7774ms |
|                  Json, small serialize |     0.0790ms |
|                Json, small deserialize |     0.1241ms |
|                  Json, large serialize |    16.3476ms |
|                Json, large deserialize |    34.5957ms |
```

vargaz committed:
Fix console bench sample.

vargaz commented:
/azp run runtime-wasm

vargaz commented:
/azp run runtime-extra-platforms

vargaz committed:
Fix landing pads.

vargaz committed:
Fix issues.

vargaz commented:
/azp run runtime-extra-platforms

vargaz commented:
/azp run runtime-extra-platforms

vargaz commented:
/azp run runtime-extra-platforms

vargaz commented:
The wasm failures are caused by emscripten OOM.

vargaz commented:
/azp run runtime-wasm

vargaz committed:
Add support for filter clauses.

vargaz commented:
With filter clauses:
```
| measurement | time |
|-:|-:|
|        Exceptions, NoExceptionHandling |     0.2323us |
|                   Exceptions, TryCatch |     0.3679us |
|              Exceptions, TryCatchThrow |     0.0163ms |
|             Exceptions, TryCatchFilter |     0.3177us |
|       Exceptions, TryCatchFilterInline |     0.2124us |
|        Exceptions, TryCatchFilterThrow |     0.0237ms |
| Exceptions, TryCatchFilterThrowApplies |     0.0201ms |
|         Json, non-ASCII text serialize |     0.4899ms |
|       Json, non-ASCII text deserialize |     1.7639ms |
|                  Json, small serialize |     0.0513ms |
|                Json, small deserialize |     0.0795ms |
|                  Json, large serialize |    14.6339ms |
|                Json, large deserialize |    22.0000ms |
```

vargaz committed:
Implement all wasm return conventions.

vargaz committed:
Fix arg/local write back.

vargaz commented:
/azp run runtime-wasm

vargaz committed:
Avoid throwing a c++ exception from do_jit_call () so the caller can clean up the interpreter stack.

vargaz commented:
/azp run runtime-wasm

vargaz commented:
/azp run runtime-wasm

vargaz commented:
/azp run runtime-wasm

vargaz committed:
Disable AOTing some more assemblies on CI.

vargaz commented:
/azp run runtime-wasm

vargaz commented:
/azp run runtime-wasm

vargaz commented:
/azp run runtime-wasm

lewing commented:
@lambdageek @BrzVlad this looks ready for review.

lambdageek commented on src/mono/mono/mini/interp/interp.c:
```diff
@@ -7433,32 +7478,38 @@ interp_run_finally_with_il_state (gpointer il_state_ptr, int clause_index, gpoin
-	clause_args.end_at_ip = (const guint16*)handler_ip_end;
-	clause_args.exit_clause = clause_index;
+	if (clause_type == MONO_EXCEPTION_CLAUSE_NONE || clause_type == MONO_EXCEPTION_CLAUSE_FILTER)
+		clause_args.end_at_ip = (const guint16*)clause_args.start_with_ip + 0xffffff;
```
Is this saying if we're in a filter or a catch clause to run until the end of the method?

vargaz replied:
Yes.

vargaz replied:
For filter clauses, it will run until ENDFILTER.

lambdageek reviewed (commented):
LGTM; but I'd like @BrzVlad 's opinion

BrzVlad commented on src/mono/mono/mini/interp/interp.c:
```diff
@@ -3897,6 +3926,8 @@ interp_exec_method (InterpFrame *frame, ThreadContext *context, FrameClauseArgs
 				error_init_reuse (error);
 				do_jit_call (context, (stackval*)(locals + return_offset), (stackval*)(locals + call_args_offset), frame, cmethod, error);
+				if (context->pending_unwind)
+					return;
```
We shouldn't simply return from `interp_exec_method`. Interpreter frames should be exited one by one. By simply returning we fail to restore the `context->data_stack`. `pending_unwind` should follow the same behavior as `has_resume_state` when it comes to exiting interpreter frames, they should probably be partly unified.

vargaz committed:
Rename llvmonly EH functions to mini_llvmonly_ for clarity.

vargaz committed:
Improve unwinding through interpreter frames.

Instead of throwing a c++ exception from mono_handle_exception ()
when an exception is caught in AOTed code, set context->has_resume_state,
so the intepreter will normally unwind until exiting interpreted code.
Then throw the c++ exception after the call to interp_exec_method ()
to unwind through the runtime code and the AOTed frames
which can't handle the exception.

vargaz commented:
/azp run runtime-wasm

vargaz commented:
/azp run runtime-extra-platforms

vargaz commented:
CI failures look unrelated.

BrzVlad commented on src/mono/mono/mini/interp/interp.c:
```diff
@@ -2641,9 +2651,21 @@ do_jit_call (ThreadContext *context, stackval *ret_sp, stackval *sp, InterpFrame
+			 * We can't rethrow here, since that will skip the cleanup of the
+			 * interpreter stack space etc. So instruct the interpreter to unwind.
+			 */
+			printf ("AAA\n");
```
left-over printf

vargaz commented:
/azp run runtime-wasm

vargaz commented:
/azp run runtime-extra-platforms

vargaz commented:
CI failures look unrelated.
